peer/ping_manager.go

```diff
@@ -131,7 +130,7 @@ func (m *PingManager) start() error {
 					e := errors.New("impossible: new ping" +
 						"in unclean state",
 					)
-					m.cfg.OnPongFailure(e)
+					go m.cfg.OnPongFailure(e)
```

Comment:
could instead leave this as is but change the actual call back implementation to do things async. Since the caller defines the call-back, it feels like the caller should be responsible for making it run async if needed

Reply:
Sorta two ways about it: the core issue here is that we made an assumption about the call back that led to the deadlock in the first place. This also isn't the first time we've run into an issue like this. Last year we had a few instances of a circular call happening in the peer during init (eg: go to fetch latest node ann, calls into server, tries to grab mutex already held) that would've been resolved by minimizing assumptions w.r.t the synchronicity of a passed call back. 

Callback aside, I think generally a better way to structure code like this would be to instead have the ping manager _signal_ another sub-system that the expectation (pong not recv'd in time) has been violated in an async manner. This way, we don't need to care about _disconnecting_ only that the _signal_ to disconnect has been sent.